### PR TITLE
Saintedlama/mongodb native

### PR DESCRIPTION
**Pull request by saintedlama**
@watson @sorribas @mafintosh Found no better place to discuss this branch. It's currently no pull request but an experimental implementation of using mongodb-native driver in the background instead of mongodb-core driver.

The reason for this is that mongodb-native implements a lot of functionality on top of mongodb-core that we would need to reimplement in mongojs too to fix our current 37 open issues. For example:

* Add support for Mongos topology #241 
* SCRAM-SHA-1 authMechanism not supported #233 
* db.close() issue #224
* SCRAM-SHA-1 missing authSource? #222
* MongoError: No more documents in tailed cursor #205

Distilled: It's auth, topologies and cursor handling that would need to be implemented as the mongodb-native driver does, which means we would need to maintain this code too!

The branch here is quite experimental and has a lot of TODOs in it. Nothing that could not be fixed but some stuff that has to be discussed.

Currently the branch suffers from two "problems":

1) cursor.explain is broken due to a bug in mongodb-core (https://github.com/christkv/mongodb-core/issues/50)
2) I'm experiencing problems with EventEmitter, Istanbul, Node.js 4.1.1 and Harmony Proxies. but this only affects coverage runs

### CODE REVIEW

**Comment by watson** [comment]
@saintedlama Wow, you've put a lot of work into this I can see :tophat: I don't have the mental capacity at the moment to weigh the pro/cons, so I'll get back on that later.

**Comment by saintedlama** [comment]
@watson thx but mostly small commits :minibus:

**Comment by saintedlama** [comment]
@watson @sorribas @mafintosh any update?

File: lib/collection.js [hunk 1]
```diff
@@ -119,17 +125,26 @@ Collection.prototype.save = function (doc, cb) {
...
     if (err) return cb(err)
-    server.remove(self._fullColName(), [{q: query, limit: justOne ? 1 : 0}], writeOpts, function (err, res) {
-      if (err) return cb(err)
-      cb(null, res.result)
-    })
+
+    // TODO: Duplicate code - see xtend below
+    if (options.justOne) {
+      collection.deleteOne(query, xtend(writeOpts, options), function (err, result) {
+        if (err) { return cb(err) }
```
Comment (mafintosh) [inline lib/collection.js]: `if (err) return cb(err)` would be fine

**Comment by mafintosh** [comment]
@saintedlama did a quick glance through your pr and overall it looks good to me.

**Comment by mafintosh** [comment]
not sure why the tests are failing though? also if @watson, @sorribas doesn't reply in the next few days feel free to merge this (assuming the tests are passing i guess)

**Comment by watson** [comment]
Forgive me if I remember this wrong, but wasn't pre v1.x using mongodb and then switched to mongodb-core in v1.x? And in that case, can someone remember the reasoning behind ditching the mongodb dependency?

**Comment by saintedlama** [comment]
@watson you remember that right :smile: I can remember that inheriting from mongodb-native was the main problem. This version does not inherit but wrap

**Comment by saintedlama** [comment]
@mafintosh this PR is not ready to merge yet but needs a bit of :heart: but I'll fix the problems when I've the :+1:

**Comment by sorribas** [comment]
@watson the latest version of `mongodb-native` at that point, broke, most APIs, so a rewrite was going to  be necessary. Because the developers of the driver recommended `mongodb-core` for "library writers",  we went down that path. Probably not the best decision.

**Comment by watson** [comment]
@sorribas thanks for explaining :smiley: 
@saintedlama I think your logic for reverting the dependency to mongodb is sound - feel free to merge when you feel it's ready :+1:

**Comment by saintedlama** [comment]
@watson @sorribas @mafintosh The failing test is a consequence of using mongodb-native. I think but I'm not quite sure that mongodb-native has some bug when it comes to explain. Already reported the issue but got a reply that everything works as designed (christkv/mongodb-core#50) - Since explain functionality is not some production critical functionality I'd prefer to adapt the tests to expect the current behavior.

@watson @sorribas I could use a little help with the pull request :smile: - just search for "TODO"

**Comment by saintedlama** [comment]
Explain bug christkv/mongodb-core#50 was fixed by commit https://github.com/christkv/mongodb-core/commit/4b98902eb97fa0712949392cc90073addf36c72b

**Comment by watson** [comment]
@saintedlama I just installed the branch into one of my projects and ran my test-suite. Completed without any hiccups :smile: :tophat: 

Well a few hiccups, but I think that's my computer doing weird stuff - the tests ran extremely slowly, but seemed to run just as slow when running with mongojd 1.4.1... weird.

Good job with all the tests btw!! Really hard work I could imagine :smile:

**Comment by mafintosh** [comment]
@saintedlama wow good job

**Comment by saintedlama** [comment]
I'll integrate the branch today in a bigger project of mine. If that runs as smooth as your test run @watson I'll merge and do a release (major, no beta)? @watson @mafintosh ok for you?

**Comment by mafintosh** [comment]
@saintedlama sounds good :+1:

**Comment by watson** [comment]
:+1:

**Comment by saintedlama** [comment]
Could add a cross platform shelljs based release script to get well done releases - ok for you? :heart: shelljs since I'm working from a windows machine :scream:

**Comment by mafintosh** [comment]
@saintedlama seems that 2.0 is working great. just wanted to say great job! if shelljs is easier for you feel free to change it

**Comment by saintedlama** [comment]
@mafintosh I'm also quite happy with 2.0 :tada: great day :beers:

**Pull request merged**